File: ios/AirMaps/AIRMapShapeOverlayManager.m
//
//  AIRMapShapeOverlayManager.m
//  AirMapsExplorer
//
//  Created by Akex Vegner on 5/17/16.
//

#import "AIRMapShapeOverlayManager.h"

#import <React/RCTConvert+CoreLocation.h>
#import <React/RCTUIManager.h>
#import <React/UIView+React.h>
#import "AIRMapShapeOverlay.h"

@interface AIRMapShapeOverlayManager () <MKMapViewDelegate>

@end

@implementation AIRMapShapeOverlayManager

RCT_EXPORT_MODULE()

- (UIView *)view {
  AIRMapShapeOverlay *overlay = [AIRMapShapeOverlay new];
  overlay.bridge = self.bridge;
  return overlay;
}

RCT_EXPORT_VIEW_PROPERTY(shapes, NSArray)
RCT_EXPORT_VIEW_PROPERTY(scale, CGFloat)
RCT_EXPORT_VIEW_PROPERTY(name, NSString)
RCT_EXPORT_VIEW_PROPERTY(rotation, NSInteger)
RCT_EXPORT_VIEW_PROPERTY(transparency, CGFloat)
RCT_EXPORT_VIEW_PROPERTY(zIndex, NSInteger)
RCT_EXPORT_VIEW_PROPERTY(onPress, RCTBubblingEventBlock)

@end

File: android/src/main/java/com/airbnb/android/react/maps/AirMapShapeOverlayManager.java
package com.airbnb.android.react.maps;

import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.common.MapBuilder;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.ViewGroupManager;
import com.facebook.react.uimanager.annotations.ReactProp;

import java.util.Map;

import javax.annotation.Nullable;

public class AirMapShapeOverlayManager extends ViewGroupManager<AirMapShapeOverlay> {

    public AirMapShapeOverlayManager() {
    }

    @Override
    public String getName() {
        return "AIRMapShapeOverlay";
    }

    @Override
    public AirMapShapeOverlay createViewInstance(ThemedReactContext context) {
        return new AirMapShapeOverlay(context);
    }

    @ReactProp(name = "shapes")
    public void setShapes(AirMapShapeOverlay view, ReadableArray shapes) {
        view.setShapes(shapes);
    }

    @ReactProp(name = "name")
    public void setOverlayName(AirMapShapeOverlay view, String name) {
        view.setOverlayName(name);
    }

    @ReactProp(name = "rotation", defaultFloat = 0.0f)
    public void setOverlayRotation(AirMapShapeOverlay view, float rotation) {
        view.setRotation(rotation);
    }

    @ReactProp(name = "scale", defaultFloat = 2.0f)
    public void setScale(AirMapShapeOverlay view, float scale) {
        view.setScale(scale);
    }

    @ReactProp(name = "transparency", defaultFloat = 0.0f)
    public void setOverlayTransparency(AirMapShapeOverlay view, float transparency) {
        view.setTransparency(transparency);
    }

    @ReactProp(name = "zIndex", defaultFloat = 1.0f)
    public void setZIndex(AirMapShapeOverlay view, float zIndex) {
        view.setZIndex(zIndex);
    }

    @Override
    @Nullable
    public Map getExportedCustomDirectEventTypeConstants() {
        return MapBuilder.of("onPress", MapBuilder.of("registrationName", "onPress"));
    }
}
